anndata/_core/aligned_mapping.py

```diff
@@ -81,6 +82,10 @@ def is_view(self) -> bool:
     def parent(self) -> Union["anndata.AnnData", "raw.Raw"]:
         return self._parent
 
+    @property
+    def parent_shape(self) -> Tuple[int, int]:
+        return self._parent.shape
+
```

Comment:
Could the shape tuple just be an attribute storing a tuple? Previously, I avoided this to reduce redundancy, but I think there's more complexity introduced by all the getters and setters. This would also remove the need for the `AnnData.__del__` method.

I was checking to see if this might break inplace subsetting, but it looks like that was already a bit broken:

```python
from anndata.tests.helpers import gen_adata

a = gen_adata((200, 100))
o1 = a.obsm
a._inplace_subset_obs(slice(50))
o2 = a.obsm
assert o1["array"].shape != o2["array"].shape
assert o1.parent is o2.parent
```

Reply:
I introduced this, for the case that the anndata object changes shape, which then should be propagated to the AxisArrays. If this cannot happen, we can drop these getters, setters and the __del__ method for sure, otherwise it should be propagated at all functions inducing different shapes explicitly. which would also be fine.

I'll have a look into this subsetting issue.

Reply:
> or it is only these _inplace_subset_* functions

The shape of the instance should only change via the `_inplace_subset_` methods.

> We can drop these getters, setters and the __del__ method for sure, when propagated at all functions inducing different shapes explicitly.

Which properties are you referring to here?

> I guess also on master it's expected that o1 is changed ?

I think it would either make sense that it change, or it no longer referred to the parent of the wrong shape.

Don't feel obligated to fix things that were already broken on master in this PR. If it's not intimately related, it's probably better to open an issue and fix it in a separate PR.

Reply:
> Which properties are you referring to here?

```python
    @property
    def parent_shape(self) -> Tuple[int, int]:
        if self._parent:
            self._parent_shape = self._parent.shape
        return self._parent_shape

    @parent_shape.setter
    def parent_shape(self, shape: Tuple[int, int]):
        self._parent_shape = shape
```
for example. When we can ensure, that the parent does not change shape other than `_inplace_subset_`, like you said, we can drop these `parent_shape` property of AxisArrays and the `__del__` method of anndata, store a tuple of the parent shape instead and update it in the `_inplace_subset_` methods when necessary.

Reply:
I updated the code to drop the `__del__` method and the `parent_shape` property.
We still need it once in the `AlignedMapping` class otherwise, a lot of subclasses has to be changed to. 
`parent_shape` does not need to be updated in `_inplace_subsets_` because they get newly created there.

Additionally I had a closer look to your solution above. 

> As an alternative, I was thinking the relationship could be reversed. The aligned mapping gets a normal reference to the anndata object, but the anndata object wraps the underlying dict whenever the attribute is accessed. This way the reference isn't circular.

However, then `_obsm` cannot be `AxisArrays`, otherwise you still have the circularity and if it's not an `AxisArrays`  it breaks a lot of other functions and would require a lot of code changes to fix it or am I wrong ?

Reply:
Sorry for the late response! I missed the question originally

I don't think many things would break if `_obsm` weren't an `AxisArray`. Anything that needs it to be should be accessing it through the public interface (whether or not they are at the moment). That said, I could be forgetting some cases.

Reply:
Sorry for my late response, too.

Yes, since creating this pull request several lines where originally `_obsm` was used now uses `obsm`.

However, when I now run the code with our idea, I still get errors that `obs_names` cannot be used for creating a view:

```python
anndata/tests/test_base.py:146: 
anndata/_core/anndata.py:848: in obs_names
    self._set_dim_index(names, "obs")
anndata/_core/anndata.py:821: in _set_dim_index
    self._init_as_actual(self.copy())
anndata/_core/anndata.py:1452: in copy
    obsm=self.obsm.copy(),
anndata/_core/anndata.py:909: in obsm
    return obsm._view(self, self.obs_names)
anndata/_core/aligned_mapping.py:92: in _view
    return self._view_class(self, parent, subset_idx)
anndata/_core/aligned_mapping.py:246: in __init__
    self.dim_names = parent_mapping.dim_names[subset_idx]
E       IndexError: arrays used as indices must be of integer (or boolean) type
```

I tried different things (using `_normalize_indices` or using `AxisArrays` directly instead of a view), but nothing resulted in a working solution. Maybe you have a idea how to fix this, since you're more familiar with the code base, but I don't have a good idea right now.
Additionally, the same changes need to be done for varm to really fix the issue, which I didn't tried with your method so far.

Reply:
I'm a little confused by what you mean here. I think I'd be able to make some suggestions if I could see the state of the code that was leading to errors. Is there a branch with these errors somewhere I can take a look at?

Reply:
That would be great. I created a branch here: https://github.com/fhausmann/anndata/tree/memory_fix_lazy_obsm

Reply:
I think I see what's happening. We try to normalize passed indices to `slice`, `int`, `array[int]`, or `array[bool]` types as soon as possible. The normalized indices are stored in view in the `_oidx` and `_vidx` attributes. `AxisArray._view` is only expecting to see those types as subset indices. This can be fixed by changing how the view is made in your `obsm` getter to:

```python
return obsm._view(self, self._oidx)
# Instead of
return obsm._view(self, self.obs_names)
```

Reply:
I fixed this, including some other small changes. However there is now another issue.
I had to introduce this https://github.com/fhausmann/anndata/blob/aa3e55b5e6251bc373a8ab94e8e11d7fc6d16dba/anndata/_core/aligned_mapping.py#L234-L237
To update the parent when `obsm` is modified. 

However there still seem to be an issue pointed out by the following test: anndata/tests/test_base.py::test_setting_dim_index[obs]
If you create an anndata object containing `obsm=dict('df'=pd.Dataframe)` , copy it and create a view it looks like as all objects are referring to the same dataframe:
`id(curr._obsm['df']) == id(orig._obsm['df']) # True` at https://github.com/theislab/anndata/blob/58886f09b2e387c6389a2de20ed0bc7d20d1b843/anndata/tests/test_base.py#L187

I think it can be fixed with creating a copy when modifying a value in `_obsm`. However this leads to infinite recursion when trying to create an anndata object from a view inplace. 

Additionally I think, these changes are now (not only) for fixing the original issue, but changing the anndata architecture. Should we now create a new pull request for this?

Reply:
Oh, the AxisArray shouldn't have it's own version of the values. That should just be a reference to the values held by the parent. That is, I would think the `__init__` should just do something like `self._data = parent._obsm`.

> If you create an anndata object containing `obsm=dict('df'=pd.Dataframe)`, copy it and create a view it looks like as all objects are referring to the same dataframe:
> `id(curr._obsm['df']) == id(orig._obsm['df']) # True at`

Is the code for this like:

```python
curr = orig[:, :].copy()
# or 
curr = orig.copy()[:, :]
```

Either way, I agree this doesn't look right. But if `curr` is a view, I'm not sure it should have values for `._obsm`. It also looks like the `.copy` method isn't actually making a copy of the dataframe if this is true, so that would be another thing to look into.

> Additionally I think, these changes are now (not only) for fixing the original issue, but changing the anndata architecture. Should we now create a new pull request for this?

I wouldn't worry too much about "architecture changes". A lot of the work I've done on this package has been to make changing the architecture easier. It's up to you how you'd like to organize this, but I often find starting a fresh PR/ branch helpful.

Reply:
Should we close this than ? Or do you think it could be worth considering if the lazy creation doesn't work out ?